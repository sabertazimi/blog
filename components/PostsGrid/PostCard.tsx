import cx from 'classnames'
import styles from './PostCard.module.css'
import Container from '@/components/Container'
import { H2 } from '@/components/Headings'
import { Calendar, Read } from '@/components/Icons'
import Link from '@/components/Link'
import { Slide } from '@/components/Motion'
import Skeleton from '@/components/Skeleton'
import { IconTag, LinkTag } from '@/components/Tags'
import type { PostMeta } from '@/types'

interface Props {
  post: PostMeta
}

function PostCard({
  post: { slug, title, createTime, readingTime, tags },
}: Props): JSX.Element {
  return (
    <Container className="card mb-16 mt-0 px-5 py-4">
      <Slide delay={0.2}>
        <LinkTag tag={tags ? tags[0] : 'Computer Science'} />
        <H2 className={styles.title}>{title}</H2>
        <IconTag
          tag={createTime ? new Date(createTime).toDateString() : 'Nowadays'}
          icon={<Calendar />}
        />
        <Container className="mt-3">
          <Skeleton
            paragraph={{ rows: Math.min(Math.floor(readingTime / 2), 10) }}
          />
          <Link
            href={`/post/${slug}`}
            className={cx(
              'flex-container float-right m-0 h-14 w-14 rounded-full border-primary',
              'bg-primary text-2xl text-light',
              'transition hover:bg-secondary hover:text-light focus:bg-secondary focus:text-light',
            )}
          >
            <Read />
          </Link>
        </Container>
      </Slide>
    </Container>
  )
}

export default PostCard
