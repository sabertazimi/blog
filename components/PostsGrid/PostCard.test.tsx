import mockData from '@mocks/data'
import { render } from '@utils'
import { axe } from 'jest-axe'
import PostCard from './PostCard'

describe('PostCard', () => {
  const mockBasePost = mockData.basePosts[0]
  const mockPost = mockData.posts[0]

  test('should render correctly (snapshot)', () => {
    const { container } = render(<PostCard post={mockPost} />)

    expect(container).toMatchSnapshot()
  })

  test('should render correctly with partial data (snapshot)', () => {
    const { container } = render(<PostCard post={mockBasePost} />)

    expect(container).toMatchSnapshot()
  })

  test('should render accessibility guidelines (AXE)', async () => {
    const { container } = render(<PostCard post={mockPost} />)

    const a11y = await axe(container, {
      rules: {
        'empty-heading': { enabled: false },
        'nested-interactive': { enabled: false },
      },
    })

    expect(a11y).toHaveNoViolations()
  })

  test('should render accessibility guidelines with partial data (AXE)', async () => {
    const { container } = render(<PostCard post={mockBasePost} />)

    const a11y = await axe(container, {
      rules: {
        'empty-heading': { enabled: false },
        'nested-interactive': { enabled: false },
      },
    })

    expect(a11y).toHaveNoViolations()
  })
})
