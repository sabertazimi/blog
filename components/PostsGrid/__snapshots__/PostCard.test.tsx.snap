// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostCard should render correctly (snapshot) 1`] = `
<div>
  <div
    class="p-auto container relative mx-auto my-0 block h-full card mb-16 mt-0 px-5 py-4"
  >
    <div
      style="opacity: 0; transform: translateX(-200px) translateZ(0);"
    >
      <span
        class="ant-tag ant-tag-has-color link dark:hover:shadow-primary"
        style="background-color: rgb(245, 159, 0);"
      >
        <a
          href="/tag/JavaScript"
        >
          JavaScript
        </a>
      </span>
      <h2
        class="ant-typography mdx-heading dark:text-light heading title h2"
      >
        1 Basic Notes
      </h2>
      <span
        class="ant-tag icon dark:border-light dark:bg-light dark:text-dark"
      >
        <span
          aria-label="calendar"
          class="anticon anticon-calendar"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="calendar"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
            />
          </svg>
        </span>
        <span>
          Wed Aug 08 2018
        </span>
      </span>
      <div
        class="p-auto container relative mx-auto my-0 block h-full mt-3"
      >
        <div
          class="ant-skeleton"
        >
          <div
            class="ant-skeleton-content"
          >
            <h3
              class="ant-skeleton-title"
              style="width: 38%;"
            />
            <ul
              class="ant-skeleton-paragraph"
            />
          </div>
        </div>
        <a
          class="flex-container float-right m-0 h-14 w-14 rounded-full border-primary bg-primary text-2xl text-light transition hover:bg-secondary hover:text-light focus:bg-secondary focus:text-light"
          href="/post/1BasicNotes"
        >
          <span
            aria-label="read"
            class="anticon anticon-read"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="read"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M928 161H699.2c-49.1 0-97.1 14.1-138.4 40.7L512 233l-48.8-31.3A255.2 255.2 0 00324.8 161H96c-17.7 0-32 14.3-32 32v568c0 17.7 14.3 32 32 32h228.8c49.1 0 97.1 14.1 138.4 40.7l44.4 28.6c1.3.8 2.8 1.3 4.3 1.3s3-.4 4.3-1.3l44.4-28.6C602 807.1 650.1 793 699.2 793H928c17.7 0 32-14.3 32-32V193c0-17.7-14.3-32-32-32zM324.8 721H136V233h188.8c35.4 0 69.8 10.1 99.5 29.2l48.8 31.3 6.9 4.5v462c-47.6-25.6-100.8-39-155.2-39zm563.2 0H699.2c-54.4 0-107.6 13.4-155.2 39V298l6.9-4.5 48.8-31.3c29.7-19.1 64.1-29.2 99.5-29.2H888v488zM396.9 361H211.1c-3.9 0-7.1 3.4-7.1 7.5v45c0 4.1 3.2 7.5 7.1 7.5h185.7c3.9 0 7.1-3.4 7.1-7.5v-45c.1-4.1-3.1-7.5-7-7.5zm223.1 7.5v45c0 4.1 3.2 7.5 7.1 7.5h185.7c3.9 0 7.1-3.4 7.1-7.5v-45c0-4.1-3.2-7.5-7.1-7.5H627.1c-3.9 0-7.1 3.4-7.1 7.5zM396.9 501H211.1c-3.9 0-7.1 3.4-7.1 7.5v45c0 4.1 3.2 7.5 7.1 7.5h185.7c3.9 0 7.1-3.4 7.1-7.5v-45c.1-4.1-3.1-7.5-7-7.5zm416 0H627.1c-3.9 0-7.1 3.4-7.1 7.5v45c0 4.1 3.2 7.5 7.1 7.5h185.7c3.9 0 7.1-3.4 7.1-7.5v-45c.1-4.1-3.1-7.5-7-7.5z"
              />
            </svg>
          </span>
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`PostCard should render correctly with partial data (snapshot) 1`] = `
<div>
  <div
    class="p-auto container relative mx-auto my-0 block h-full card mb-16 mt-0 px-5 py-4"
  >
    <div
      style="opacity: 0; transform: translateX(-200px) translateZ(0);"
    >
      <span
        class="ant-tag ant-tag-has-color link dark:hover:shadow-primary"
        style="background-color: rgb(66, 99, 235);"
      >
        <a
          href="/tag/Computer Science"
        >
          Computer Science
        </a>
      </span>
      <h2
        class="ant-typography mdx-heading dark:text-light heading title h2"
      >
        1 Basic Notes
      </h2>
      <span
        class="ant-tag icon dark:border-light dark:bg-light dark:text-dark"
      >
        <span
          aria-label="calendar"
          class="anticon anticon-calendar"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="calendar"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
            />
          </svg>
        </span>
        <span>
          Nowadays
        </span>
      </span>
      <div
        class="p-auto container relative mx-auto my-0 block h-full mt-3"
      >
        <div
          class="ant-skeleton"
        >
          <div
            class="ant-skeleton-content"
          >
            <h3
              class="ant-skeleton-title"
              style="width: 38%;"
            />
            <ul
              class="ant-skeleton-paragraph"
            />
          </div>
        </div>
        <a
          class="flex-container float-right m-0 h-14 w-14 rounded-full border-primary bg-primary text-2xl text-light transition hover:bg-secondary hover:text-light focus:bg-secondary focus:text-light"
          href="/post/1BasicNotes"
        >
          <span
            aria-label="read"
            class="anticon anticon-read"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="read"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M928 161H699.2c-49.1 0-97.1 14.1-138.4 40.7L512 233l-48.8-31.3A255.2 255.2 0 00324.8 161H96c-17.7 0-32 14.3-32 32v568c0 17.7 14.3 32 32 32h228.8c49.1 0 97.1 14.1 138.4 40.7l44.4 28.6c1.3.8 2.8 1.3 4.3 1.3s3-.4 4.3-1.3l44.4-28.6C602 807.1 650.1 793 699.2 793H928c17.7 0 32-14.3 32-32V193c0-17.7-14.3-32-32-32zM324.8 721H136V233h188.8c35.4 0 69.8 10.1 99.5 29.2l48.8 31.3 6.9 4.5v462c-47.6-25.6-100.8-39-155.2-39zm563.2 0H699.2c-54.4 0-107.6 13.4-155.2 39V298l6.9-4.5 48.8-31.3c29.7-19.1 64.1-29.2 99.5-29.2H888v488zM396.9 361H211.1c-3.9 0-7.1 3.4-7.1 7.5v45c0 4.1 3.2 7.5 7.1 7.5h185.7c3.9 0 7.1-3.4 7.1-7.5v-45c.1-4.1-3.1-7.5-7-7.5zm223.1 7.5v45c0 4.1 3.2 7.5 7.1 7.5h185.7c3.9 0 7.1-3.4 7.1-7.5v-45c0-4.1-3.2-7.5-7.1-7.5H627.1c-3.9 0-7.1 3.4-7.1 7.5zM396.9 501H211.1c-3.9 0-7.1 3.4-7.1 7.5v45c0 4.1 3.2 7.5 7.1 7.5h185.7c3.9 0 7.1-3.4 7.1-7.5v-45c.1-4.1-3.1-7.5-7-7.5zm416 0H627.1c-3.9 0-7.1 3.4-7.1 7.5v45c0 4.1 3.2 7.5 7.1 7.5h185.7c3.9 0 7.1-3.4 7.1-7.5v-45c.1-4.1-3.1-7.5-7-7.5z"
              />
            </svg>
          </span>
        </a>
      </div>
    </div>
  </div>
</div>
`;
